File: src/main/java/com/epicsquid/whoosh/cap/TransporterCap.java
package com.epicsquid.whoosh.cap;

import com.epicsquid.whoosh.utils.TransporterPoint;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.nbt.Tag;

import java.util.HashMap;
import java.util.Map;

public class TransporterCap implements ITransporterCap{

	private final Map<String, TransporterPoint> points = new HashMap<>();

	@Override
	public Map<String, TransporterPoint> getPoints() {
		return Map.copyOf(points);
	}

	@Override
	public void addPoint(String name, TransporterPoint point) {
		points.put(name, point);
	}

	@Override
	public void removePoint(String name) {
		points.remove(name);
	}

	@Override
	public TransporterPoint getPoint(String name) {
		return points.get(name);
	}

	@Override
	public CompoundTag serializeNBT() {
		CompoundTag tag = new CompoundTag();
		for (String key : points.keySet()) {
			tag.put(key, points.get(key).serializeNBT());
		}
		return tag;
	}

	@Override
	public void deserializeNBT(CompoundTag nbt) {
		for (String key : nbt.getAllKeys()) {
			Tag tag = nbt.get(key);
			if (tag instanceof CompoundTag ct) {
				points.put(key, TransporterPoint.fromNBT(ct));
			}
		}
	}
}

File: src/main/java/dev/epicsquid/whoosh/cap/TransporterCap.kt
package dev.epicsquid.whoosh.cap

import dev.epicsquid.whoosh.utils.TransporterPoint
import net.minecraft.nbt.CompoundTag

object TransporterCap : ITransporterCap {
    override val points: MutableMap<String?, TransporterPoint?> = HashMap()
    fun getPoints(): Map<String?, TransporterPoint?> {
        return java.util.Map.copyOf(points)
    }

    override fun addPoint(name: String?, point: TransporterPoint?) {
        points[name] = point
    }

    override fun removePoint(name: String?) {
        points.remove(name)
    }

    override fun getPoint(name: String?): TransporterPoint? {
        return points[name]
    }

    override fun serializeNBT(): CompoundTag? {
        val tag = CompoundTag()
        for (key in points.keys) {
            tag.put(key, points[key]!!.serializeNBT())
        }
        return tag
    }

    override fun deserializeNBT(nbt: CompoundTag?) {
//        for (key in nbt.allKeys) {
//            val tag = nbt[key]
//            if (tag is CompoundTag) {
//                points[key] = TransporterPoint.fromNBT(tag)
//            }
//        }
    }
}
